File: X-Native/apps/web/src/engine/transaction.ts
/**
 * X-Native Canonical Command/Transaction System (Phase 0 Foundation)
 *
 * All document mutations flow through explicit, atomic Transactions composed
 * of reversible Operations. This guarantees:
 * 1. Deterministic local Undo/Redo by inverting operations.
 * 2. Immutable transaction log stream for future Phase 6 CRDT sync.
 * 3. Atomic multi-property or multi-node mutations with rollback on failure.
 * 4. Extensibility for Plugin API and automation scripting.
 */

import type { XNode, VectorNetwork } from "./types";
import type { Modifier } from "./modifierStack";

export type TransactionId = string;

export interface BaseOperation {
  type: string;
}

export interface SetPropertyOp extends BaseOperation {
  type: "setProperty";
  targetId: string;
  property: string;
  oldValue: any;
  newValue: any;
}

export interface InsertNodeOp extends BaseOperation {
  type: "insertNode";
  parentId: string;
  node: XNode;
  index?: number;
}

export interface RemoveNodeOp extends BaseOperation {
  type: "removeNode";
  parentId: string;
  nodeId: string;
  previousIndex: number;
  previousNode: XNode;
}

export interface MoveNodeOp extends BaseOperation {
  type: "moveNode";
  nodeId: string;
  oldParentId: string;
  newParentId: string;
  oldIndex: number;
  newIndex: number;
}

export interface SetVariableOp extends BaseOperation {
  type: "setVariable";
  variableId: string;
  oldValue: any;
  newValue: any;
}

export interface SetVectorNetworkOp extends BaseOperation {
  type: "setVectorNetwork";
  targetId: string;
  oldNetwork?: VectorNetwork;
  newNetwork: VectorNetwork;
}

export interface ApplyModifierOp extends BaseOperation {
  type: "applyModifier";
  targetId: string;
  modifier: Modifier;
  index?: number;
}

export interface RemoveModifierOp extends BaseOperation {
  type: "removeModifier";
  targetId: string;
  index: number;
  previousModifier: Modifier;
}

export interface SetExpressionOp extends BaseOperation {
  type: "setExpression";
  targetId: string;
  property: string;
  oldExpr?: string;
  newExpr: string;
}

export type Operation =
  | SetPropertyOp
  | InsertNodeOp
  | RemoveNodeOp
  | MoveNodeOp
  | SetVariableOp
  | SetVectorNetworkOp
  | ApplyModifierOp
  | RemoveModifierOp
  | SetExpressionOp;

export interface TransactionMetadata {
  label?: string;
  authorId?: string;
  origin?: "user" | "script" | "plugin" | "history" | "sync";
}

export interface Transaction {
  id: TransactionId;
  timestamp: number;
  operations: Operation[];
  metadata?: TransactionMetadata;
}

/**
 * Creates a unique transaction id with high-resolution timestamp.
 */
let txCounter = 0;
export function createTransactionId(): TransactionId {
  txCounter += 1;
  return `tx_${Date.now()}_${txCounter.toString(36)}_${Math.random().toString(36).slice(2, 7)}`;
}

/**
 * Computes the exact mathematical inverse of an Operation.
 */
export function invertOperation(op: Operation): Operation {
  switch (op.type) {
    case "setProperty":
      return {
        type: "setProperty",
        targetId: op.targetId,
        property: op.property,
        oldValue: op.newValue,
        newValue: op.oldValue,
      };

    case "insertNode":
      return {
        type: "removeNode",
        parentId: op.parentId,
        nodeId: op.node.id,
        previousIndex: op.index ?? 0,
        previousNode: JSON.parse(JSON.stringify(op.node)),
      };

    case "removeNode":
      return {
        type: "insertNode",
        parentId: op.parentId,
        node: JSON.parse(JSON.stringify(op.previousNode)),
        index: op.previousIndex,
      };

    case "moveNode":
      return {
        type: "moveNode",
        nodeId: op.nodeId,
        oldParentId: op.newParentId,
        newParentId: op.oldParentId,
        oldIndex: op.newIndex,
        newIndex: op.oldIndex,
      };

    case "setVariable":
      return {
        type: "setVariable",
        variableId: op.variableId,
        oldValue: op.newValue,
        newValue: op.oldValue,
      };

    case "setVectorNetwork":
      return {
        type: "setVectorNetwork",
        targetId: op.targetId,
        oldNetwork: JSON.parse(JSON.stringify(op.newNetwork)),
        newNetwork: op.oldNetwork ? JSON.parse(JSON.stringify(op.oldNetwork)) : { vertices: [], segments: [] },
      };

    case "applyModifier":
      return {
        type: "removeModifier",
        targetId: op.targetId,
        index: op.index ?? 0,
        previousModifier: JSON.parse(JSON.stringify(op.modifier)),
      };

    case "removeModifier":
      return {
        type: "applyModifier",
        targetId: op.targetId,
        modifier: JSON.parse(JSON.stringify(op.previousModifier)),
        index: op.index,
      };

    case "setExpression":
      return {
        type: "setExpression",
        targetId: op.targetId,
        property: op.property,
        oldExpr: op.newExpr,
        newExpr: op.oldExpr ?? "",
      };

    default:
      throw new Error(`Cannot invert unknown operation type: ${(op as any).type}`);
  }
}

/**
 * Computes the exact mathematical inverse of a Transaction.
 * Inverting a transaction reverses the operations and inverts each one.
 */
export function invertTransaction(tx: Transaction): Transaction {
  const invertedOps: Operation[] = [];
  for (let i = tx.operations.length - 1; i >= 0; i--) {
    invertedOps.push(invertOperation(tx.operations[i]));
  }
  return {
    id: createTransactionId(),
    timestamp: Date.now(),
    operations: invertedOps,
    metadata: {
      ...tx.metadata,
      label: tx.metadata?.label ? `Undo: ${tx.metadata.label}` : "Inverted Transaction",
      origin: "history",
    },
  };
}

/**
 * TransactionStream manages undo/redo history and provides delta streaming
 * to subscribers (such as future CRDT engines or plugin loggers).
 */
export class TransactionStream {
  private undoStack: Transaction[] = [];
  private redoStack: Transaction[] = [];
  private subscribers: Set<(tx: Transaction) => void> = new Set();
  private maxHistory: number;

  constructor(maxHistory: number = 200) {
    this.maxHistory = maxHistory;
  }

  /**
   * Records and emits an executed transaction into the stream.
   * Clears redo stack unless originated from history.
   */
  public push(tx: Transaction): void {
    if (tx.operations.length === 0) return;

    if (tx.metadata?.origin !== "history") {
      this.redoStack.length = 0;
      this.undoStack.push(tx);
      if (this.undoStack.length > this.maxHistory) {
        this.undoStack.shift();
      }
    }

    // Broadcast to delta stream subscribers
    for (const sub of this.subscribers) {
      try {
        sub(tx);
      } catch (err) {
        console.error("Error in transaction subscriber:", err);
      }
    }
  }

  /**
   * Prepares undo: pops the last transaction and returns its inverted counterpart.
   */
  public popUndo(): Transaction | null {
    const tx = this.undoStack.pop();
    if (!tx) return null;
    const inv = invertTransaction(tx);
    this.redoStack.push(tx);
    return inv;
  }

  /**
   * Prepares redo: pops the last redo transaction and returns it for re-application.
   */
  public popRedo(): Transaction | null {
    const tx = this.redoStack.pop();
    if (!tx) return null;
    this.undoStack.push(tx);
    return tx;
  }

  public canUndo(): boolean {
    return this.undoStack.length > 0;
  }

  public canRedo(): boolean {
    return this.redoStack.length > 0;
  }

  public subscribe(callback: (tx: Transaction) => void): () => void {
    this.subscribers.add(callback);
    return () => {
      this.subscribers.delete(callback);
    };
  }

  public clear(): void {
    this.undoStack.length = 0;
    this.redoStack.length = 0;
  }

  public getHistory(): { undoCount: number; redoCount: number } {
    return {
      undoCount: this.undoStack.length,
      redoCount: this.redoStack.length,
    };
  }
}

File: X-Native/crates/x-core/src/transaction.rs
//! Canonical Command/Transaction System (Phase 0 Foundation)
//!
//! Every document mutation flows through explicit, atomic Transactions composed
//! of reversible Operations. This guarantees:
//! 1. Deterministic local Undo/Redo by inverting operations.
//! 2. Immutable transaction delta stream for Phase 6 CRDT sync.
//! 3. Single source of truth for Editor, Renderer, and Prototype runtime.

use serde::{Deserialize, Serialize};

pub type TransactionId = String;
pub type NodeId = String;
pub type VariableId = String;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum Operation {
    SetProperty {
        target_id: NodeId,
        property: String,
        old_value: String,
        new_value: String,
    },
    InsertNode {
        parent_id: NodeId,
        node_id: NodeId,
        index: usize,
    },
    RemoveNode {
        parent_id: NodeId,
        node_id: NodeId,
        previous_index: usize,
    },
    MoveNode {
        node_id: NodeId,
        old_parent_id: NodeId,
        new_parent_id: NodeId,
        old_index: usize,
        new_index: usize,
    },
    SetVariable {
        variable_id: VariableId,
        old_value: String,
        new_value: String,
    },
    DeleteEdge {
        edge_id: String,
    },
    ApplyModifier {
        target_id: NodeId,
        modifier_json: String,
        index: usize,
    },
    RemoveModifier {
        target_id: NodeId,
        index: usize,
        previous_modifier_json: String,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct TransactionMetadata {
    pub label: Option<String>,
    pub author_id: Option<String>,
    pub origin: Option<String>, // "user", "script", "plugin", "sync"
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Transaction {
    pub id: TransactionId,
    pub timestamp: u64,
    pub operations: Vec<Operation>,
    pub metadata: Option<TransactionMetadata>,
}

impl Operation {
    /// Inverts the operation mathematically for deterministic undo.
    pub fn invert(&self) -> Self {
        match self {
            Operation::SetProperty {
                target_id,
                property,
                old_value,
                new_value,
            } => Operation::SetProperty {
                target_id: target_id.clone(),
                property: property.clone(),
                old_value: new_value.clone(),
                new_value: old_value.clone(),
            },
            Operation::InsertNode {
                parent_id,
                node_id,
                index,
            } => Operation::RemoveNode {
                parent_id: parent_id.clone(),
                node_id: node_id.clone(),
                previous_index: *index,
            },
            Operation::RemoveNode {
                parent_id,
                node_id,
                previous_index,
            } => Operation::InsertNode {
                parent_id: parent_id.clone(),
                node_id: node_id.clone(),
                index: *previous_index,
            },
            Operation::MoveNode {
                node_id,
                old_parent_id,
                new_parent_id,
                old_index,
                new_index,
            } => Operation::MoveNode {
                node_id: node_id.clone(),
                old_parent_id: new_parent_id.clone(),
                new_parent_id: old_parent_id.clone(),
                old_index: *new_index,
                new_index: *old_index,
            },
            Operation::SetVariable {
                variable_id,
                old_value,
                new_value,
            } => Operation::SetVariable {
                variable_id: variable_id.clone(),
                old_value: new_value.clone(),
                new_value: old_value.clone(),
            },
            Operation::DeleteEdge { edge_id } => Operation::DeleteEdge {
                edge_id: edge_id.clone(),
            },
            Operation::ApplyModifier {
                target_id,
                modifier_json,
                index,
            } => Operation::RemoveModifier {
                target_id: target_id.clone(),
                index: *index,
                previous_modifier_json: modifier_json.clone(),
            },
            Operation::RemoveModifier {
                target_id,
                index,
                previous_modifier_json,
            } => Operation::ApplyModifier {
                target_id: target_id.clone(),
                modifier_json: previous_modifier_json.clone(),
                index: *index,
            },
        }
    }
}

impl Transaction {
    /// Inverts the entire transaction.
    pub fn invert(&self) -> Self {
        let mut inverted_ops: Vec<Operation> = self
            .operations
            .iter()
            .rev()
            .map(|op| op.invert())
            .collect();

        Self {
            id: format!("inv_{}", self.id),
            timestamp: self.timestamp,
            operations: inverted_ops,
            metadata: Some(TransactionMetadata {
                label: self.metadata.as_ref().and_then(|m| m.label.as_ref()).map(|l| format!("Undo: {}", l)),
                author_id: self.metadata.as_ref().and_then(|m| m.author_id.clone()),
                origin: Some("history".to_string()),
            }),
        }
    }
}
